kubectl-volsync/cmd/migration_create.go

```diff
@@ -72,34 +104,329 @@ func validateMigrationCreate(cmd *cobra.Command, args []string) error {
+		v1.ServiceType(serviceType) != v1.ServiceTypeLoadBalancer {
+		return fmt.Errorf("unsupported service type: %v", v1.ServiceType(serviceType))
+	}
+	mrd.Destination.ServiceType = (*v1.ServiceType)(&serviceType)
+	mrd.MDName = mrd.Namespace + "-" + mrd.PVCName + "-migration-dest"
+
+	return nil
+}
+
+func (mc *migrationCreate) Run(ctx context.Context) error {
+	// create namespace if does not exist
+	err := mc.createNamespace(ctx)
 	if err != nil {
 		return err
 	}
-	rName, err := cmd.Flags().GetString("relationship")
+	// create destination PVC if does not exist
+	if mc.PVC == nil {
+		mc.PVC, err = mc.createDestinationPVC(ctx)
+		if err != nil {
+			return err
+		}
+	}
+	// create migration destination
+	err = mc.createDestination(ctx)
 	if err != nil {
 		return err
 	}
-	relation, err := createRelationship(configDir, rName, MigrationRelationship)
+	// save the destination details into config file
+	if err = mc.mr.Save(); err != nil {
+		return fmt.Errorf("unable to save relationship configuration: %w", err)
+	}
+	return nil
+}
+
+func (mc *migrationCreate) createNamespace(ctx context.Context) error {
+	mrd := mc.mr.data.Destination
+	ns := &v1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: mrd.Namespace,
+		},
+	}
+	if err := mc.clientObject.Create(ctx, ns); err != nil {
+		if kerrs.IsAlreadyExists(err) {
+			klog.Infof("Namespace: \"%s\" already present, proceeding with this namespace",
+				mrd.Namespace)
+			return nil
+		}
+		return err
+	}
+	klog.Infof("Created destination namespace: \"%s\"", mrd.Namespace)
+
+	return nil
+}
+
+func (mc *migrationCreate) createDestinationPVC(ctx context.Context) (*v1.PersistentVolumeClaim, error) {
+	mrd := mc.mr.data.Destination
+	var destPVC *v1.PersistentVolumeClaim
+
+	if *mrd.Destination.StorageClassName == "" {
+		destPVC = &v1.PersistentVolumeClaim{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      mrd.PVCName,
+				Namespace: mrd.Namespace,
+			},
+			Spec: v1.PersistentVolumeClaimSpec{
+				AccessModes: mrd.Destination.AccessModes,
+				Resources: v1.ResourceRequirements{
+					Requests: v1.ResourceList{
+						v1.ResourceStorage: *mrd.Destination.Capacity,
+					},
+				},
+			},
+		}
+	} else {
+		destPVC = &v1.PersistentVolumeClaim{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      mrd.PVCName,
+				Namespace: mrd.Namespace,
+			},
+			Spec: v1.PersistentVolumeClaimSpec{
+				AccessModes:      mrd.Destination.AccessModes,
+				StorageClassName: mrd.Destination.StorageClassName,
+				Resources: v1.ResourceRequirements{
+					Requests: v1.ResourceList{
+						v1.ResourceStorage: *mrd.Destination.Capacity,
+					},
+				},
+			},
+		}
+	}
+	if err := mc.clientObject.Create(ctx, destPVC); err != nil {
+		return nil, err
+	}
+	mrd.Destination.StorageClassName = destPVC.Spec.StorageClassName
+	klog.Infof("Created destination PVC: \"%s\"", mrd.PVCName)
+
+	return destPVC, nil
+}
+
+func (mc *migrationCreate) deleteDestinationPVC(ctx context.Context) error {
```

Comment:
It looks like this is only here for the tests?
The tests should be structured to clean themselves up. Take a look at
https://github.com/backube/volsync/blob/main/controllers/mover/restic/restic_test.go#L220-L228
and
https://github.com/backube/volsync/blob/main/controllers/mover/restic/restic_test.go#L272-L275
that set up and clean up from each test. The setup creates a new Namespace in which the test can run, and the cleanup deletes that NS, causing all the (namespaced) objects created during the test to be deleted.